storage/libradb/src/ledger_store/mod.rs

```diff
@@ -192,6 +213,14 @@ impl LedgerStore {
         ledger_info_with_sigs: &LedgerInfoWithSignatures,
         cs: &mut ChangeSet,
     ) -> Result<()> {
+        let ledger_info = ledger_info_with_sigs.ledger_info();
+
+        if ledger_info.next_validator_set().is_some() {
+            let epoch = ledger_info_with_sigs.ledger_info().epoch();
+            let version = ledger_info_with_sigs.ledger_info().version();
+            cs.batch
+                .put::<EpochByVersionSchema>(&(version + 1), &(epoch + 1))?;
```

Comment:
@zekun000  Just to confirm, once a LedgerInfo with a ValidatorSet is commited, it's guaranteed that next version is in a new epoch, right?

Reply:
yeah, but it's not guaranteed next LI is for next version though(at least for now), is it a problem?

there could be a version 5 that has LI{epoch: 1, has_validators} and then we commit another empty block and write LI{epoch: 2} for version 5

Reply:
@zekun000  Wait, you are sending two LI's with version epoch of 1 & 2 respectively, both for version 5? I hope it's a typo >_<

Reply:
it's not lol, like today when we commit empty blocks we'll overwrite the ledger info for the same epoch.

i hope this'll be addressed by the block prologue transaction so that there's no more empty blocks.

but for now yeah, we may commit two LI with epoch 1&2 for the same version 5(well at most 2 epochs since reconfiguration must be a txn if that makes you feel better), and I think version 5 should be mapped to epoch 2, if clients knows about this txn, it should move the epoch 2 already.

Reply:
@zekun000  Let's summarize: 

You send LI51(v5, e0, Some(next_validator_set)).
-> I write down the LI as final ledger_info for epoch 0.
-> I write down index item (v5 -> e1). When you get_epoch(5) I return epoch 1.

You send LI52 with version 5 + epoch 1 + None.
I write down the LI as first ledger_info for epoch 1 ( to be overwritten by the next transaction).
When you get_latest_ledger_infos_per_epoch(since:0), I return [LI51, LI52]; to elaborate, that's [(v5, e0, Some(next_validator_set)), (v5, e1, None)] (Note that you have two v5's)

You send LI61 with version 6 + epoch 1 + None
When you get_latest_ledger_infos_per_epoch(since:0), I return [LI51, LI61]; to elaborate, that's [(v5, e0, Some(next_validator_set)), (v6, e1, None)]

^ If this is expected, I'll update this line to `put(&version, &(epoch+1)))`.
 
Also how is the initial validator_set written? Does the genesis LI carry Some(next_validator_set)? do we start with epoch 0 or epoch 1?

Reply:
yeah, that's the expected behavior, also for `update_to_latest_ledger`, if client version is 4, we sent back [LI51], if client version is 5, we sent back [] as ValidatorChangeEventWithProof.

for genesis LI, we're supposed to parse the validator set from executed result see https://github.com/libra/libra/blob/master/executor/src/lib.rs#L387

this LI is at epoch 0, so genesis txn is mapped to epoch 1 iiuc.